Feature/conform

john-paulsmith opened the pull request:
This change adds new properties and an action to help hosts with editorial timeline conform, as described in #217.

`kOfxPropAka` will allow plug-ins to list alternative names by which any named object could be known. A `kOfxPropAka2D` property covers multidimensional parameters. For example, an effect could list the name it is known by in a different API such as AVX or Adobe. It can also list the names its parameters are known by, which might be GUIDs or similar.

The conform action, `kOfxActionConform`, will allow plug-ins to update their parameter values and keyframes based on information passed from the host. Prior to invoking this action, hosts will set parameter values to the best of their ability, and the plug-in will examine these and update them based on internal knowledge about where they came from. For example, a spatial parameter might need to be adjusted to compensate for different origin.

john-paulsmith committed:
Alternative names for any object

A host may be creating an effect instance based on information imported from another host which uses a different plug-in API. In this case, effects and parameters might have different names. This property allows a plug-in to specify alternative names for any object such as effects and parameters.

Signed-off-by: John-Paul Smith <jps@borisfx.com>

john-paulsmith committed:
Conform action

This new action allows plug-ins to conform their parameters to match a project imported from a non-OpenFX host.

Signed-off-by: John-Paul Smith <jps@borisfx.com>

john-paulsmith committed:
Alternative names for multidimensional params

Some foreign APIs don’t support multidimensional parameters. Plug-ins can set kOfxPropAka2D to provide alternative names for each dimenesion of a parameter, allowing the host to map them correctly.

Signed-off-by: John-Paul Smith <jps@borisfx.com>

barretpj commented on include/ofxCore.h:
```diff
@@ -705,6 +705,15 @@ This property is used to provide alternative names for objects. It can be used i
 
+/** @brief Alternative names for parameter dimensions.
+
+    - Type - string X N X M
```
This is, I assume, for the case where the OFX plugin uses a single parameter with a 2D param type, but one foreign host uses (e.g.) `TranslateX` and `TranslateY` and another uses `tx` and `ty`. How do you store those mappings in a 1D string array property? What does "string X N X M" mean?

john-paulsmith committed:
Simpler approach for multidimensional parameters

The previous approach to alternative names for multidimensional parameters didn’t define how the array was indexed. Because the number of dimensions in our known parameter types is fixed and has a maximum of 4, a simpler approach of hard-coded properties can be used instead.

Signed-off-by: John-Paul Smith <jps@borisfx.com>
